### ISSUE TITLE
Testing patterns: catalogue, aiming for simplifications

### ISSUE THREAD

**Issue by jiridanek** [opendatahub-io/notebooks#3433]
_A catalogue of all testing patterns in the notebooks repository, covering papermill-on-OpenShift, testcontainers with exec/injection, K8s API integration, browser/Playwright, static checks, unit tests, Makefile validate-* targets, Go tests, CI linting/scanning, and infra smoke workflows._

# Testing Patterns Catalogue

This repo has **eleven distinct testing patterns**, each with different execution environments, entry points, and verification strategies.

---

## 1. Papermill on OpenShift (shell-driven e2e)

**Entry point:** `make test-<target>` -> [`scripts/test_jupyter_with_papermill.sh`](scripts/test_jupyter_with_papermill.sh)

**Lifecycle:** Three separate `make` targets, run in sequence (not chained):

```
make deploy9-<target>     # kubectl apply -k (kustomize)
make test-<target>        # papermill in-pod
make undeploy9-<target>   # kubectl delete -k
```

**How it works:**

- Assumes a pod is already running from `deploy9-*`
- Waits for pod readiness, sets up `kubectl port-forward` to `:8888`
- Installs papermill in-pod: `python3 -m pip install papermill`
- `kubectl cp`s the `jupyter/<flavor>/.../test/` directory into the pod
- Builds `expected_versions.json` from imagestream manifests + requirements files
- Runs: `kubectl exec <pod> -- python3 -m papermill test_notebook.ipynb output.ipynb`
- Assertion: `grep FAILED` on stderr output file
- For datascience-derived images, runs minimal notebook first, then datascience, then flavor-specific

**Test notebooks:** 8 files at `jupyter/<flavor>/ubi9-python-3.12/test/test_notebook.ipynb`

**Configuration:**
- `KONFLUX=yes|no` selects manifest path (`manifests/rhoai/base/` vs `manifests/odh/base/`)
- `IMAGE_TAG`, `IMAGE_REGISTRY` set at deploy time
- Pod is selected by `app=<notebook_name>` label

**Key files:**
- [`scripts/test_jupyter_with_papermill.sh`](scripts/test_jupyter_with_papermill.sh) -- sole driver script
- [`tests/manifests.py`](tests/manifests.py) -- Python parity for the shell path-resolution logic
- [`Makefile`](Makefile) (lines ~233-300) -- `deploy9-%`, `test-%`, `undeploy9-%`

---

## 2. Testcontainers (Python, local containers)

**Entry point:** `pytest tests/containers --image=<image>` (excluded from default collection via `tests/conftest.py`)

**How containers are started:**

- **Sleep-infinity pattern** (most tests): `DockerContainer(image).with_command("/bin/sh -c 'sleep infinity'")` -- container stays alive while tests `exec` into it
- **Real-startup pattern** (`WorkbenchContainer`): subclasses `DockerContainer`, uses image default CMD, `tty=True`, `tmpfs` for `/opt/app-root/src`, exposed port 8888, HTTP readiness check
- **Networked pattern**: `MySqlContainer` + `Network` for multi-container topologies (e.g., datascience MySQL tests)

**Session setup** ([`tests/containers/conftest.py`](tests/containers/conftest.py)):
- Configures Ryuk (privileged mode for Podman/SELinux)
- Resolves Docker socket path (host vs Podman machine)
- `--image` CLI option, repeatable; parametrizes via `pytest_generate_tests`
- Label-based skip logic: image labels determine which fixture/test applies (`workbench_image`, `runtime_image`, `cuda_image`, etc.)

**Test layers:**
- [`base_image_test.py`](tests/containers/base_image_test.py) -- ELF linking, CLI tools, pip, env vars
- [`runtime_test.py`](tests/containers/runtimes/runtime_test.py) -- pyzmq, mlflow imports
- [`workbench_image_test.py`](tests/containers/workbenches/workbench_image_test.py) -- HTTP readiness, IPv6
- [`manifest_validation_test.py`](tests/containers/manifest_validation_test.py) -- `pip list` vs manifest versions
- Flavor-specific: JupyterLab, RStudio, GPU library loading

**Utilities** ([`docker_utils.py`](tests/containers/docker_utils.py)):
- `container_cp` / `from_container_cp` -- tar-based file transfer
- `container_exec` / `container_exec_with_stdin` -- streaming exec with TTY/stdin support

---

## 3. Python injection patterns (exec + code transport)

Three distinct strategies for running Python inside containers:

### 3a. Base64-encoded function transport

[`base_image_test.py`](tests/containers/base_image_test.py) defines `encode_python_function_execution_command_interpreter()`:

- Uses `inspect.getsource(function)` to grab a Python function's source
- Base64-encodes it, then generates a `python -c` command:
  `python -c "import binascii; s=binascii.a2b_base64('...'); exec(s.decode()); print(func_name(args))"`
- Used for ELF/ldd checks, pyzmq/mlflow imports, GPU library loading
- GPU variant in [`gpu_library_loading_test.py`](tests/containers/workbenches/gpu_library_loading_test.py) outputs `RESULT>` + JSON for structured assertions

### 3b. Inline `python -c` with string blobs

- [`jupyterlab_datascience_test.py`](tests/containers/workbenches/jupyterlab/jupyterlab_datascience_test.py): multiline f-string SQL/SASL setup code passed directly to `[python_exe, "-c", code_string]`
- [`accelerator_image_test.py`](tests/containers/workbenches/accelerator_image_test.py): `shlex.join(["python", "-c", torch_check])` for GPU verification (over K8s exec, not testcontainers)

### 3c. File-based execution

- Host writes a script, `container_cp` copies it into container, then `container.exec(["python", "/path/to/script.py"])`
- Used for sklearn smoke tests, TrustyAI checks, `libraries_testunits.py`

---

## 4. Kubernetes / OpenShift API tests (Python client)

**Entry point:** `pytest tests/containers -m openshift --image=<image>`

**Client libraries:**
- `kubernetes` (official) -- `CoreV1Api`, `stream` (exec, portforward), dynamic client
- `openshift-python-wrapper` (`ocp_resources`) -- `get_client()`, `Deployment`, `PVC`, `Pod`, `Project`/`Namespace`

**Central helper:** [`kubernetes_utils.py`](tests/containers/kubernetes_utils.py)
- `ImageDeployment` context manager: creates test namespace, PVC (1Gi RWO), Deployment with `restricted-v2` SCC, waits for pod, in-process port-forward to 8888
- `exec()` via `kubernetes.stream` WebSocket
- Auth: delegates to `ocp_resources.resource.get_client()` (kubeconfig / in-cluster)
- Username discovery: `SelfSubjectReview` API (`authentication.k8s.io/v1`)
- Cleanup: `TestFrame` deferred stack calls `resource.clean_up()`

**Test files using this:**
- [`workbench_image_test.py`](tests/containers/workbenches/workbench_image_test.py) -- `test_image_run_on_openshift`
- [`accelerator_image_test.py`](tests/containers/workbenches/accelerator_image_test.py) -- CUDA/ROCm on OpenShift with GPU resource requests

**CI replay (shell subprocess):** [`ci/cached-builds/make_test.py`](ci/cached-builds/make_test.py) -- `kubectl create namespace`, `make deploy9-*`, wait, `make test-*`, undeploy

---

## 5. Browser tests (Playwright)

**Entry point:** `tests/browser/` with `npx playwright test`

**Two flavors:**

### 5a. Code-server (Testcontainers + Playwright)

[`codeserver.spec.ts`](tests/browser/tests/codeserver.spec.ts):
- **Node** `testcontainers` (`GenericContainer`) starts code-server image
- `HttpWaitStrategy` for readiness
- Playwright connects via CDP to exercise the UI
- No OpenShift dependency

### 5b. OpenShift console

[`openshift_console.spec.ts`](tests/browser/tests/openshift_console.spec.ts):
- Uses `@kubernetes/client-node` to read the `console` Route from `openshift-console` namespace
- Playwright navigates to the discovered HTTPS URL
- Tagged `@openshift`; requires `KUBECONFIG` env var

**Config:** [`playwright.config.ts`](tests/browser/playwright.config.ts) -- projects, `DEFAULT_TEST_IMAGE`, JUnit output, traces on retry

---

## 6. Static and unit tests (default pytest)

**Entry point:** `make test` / `pytest` (default collection: `tests/` + `ntb/`, excluding `tests/containers/`)

### 6a. Static / policy tests

[`tests/test_main.py`](tests/test_main.py) -- the bulk of default-run validation:
- Dockerfile `subscription-manager` pattern guard
- `pyproject.toml` + `pylock.toml` consistency (versions, markers, specifiers)
- Manifest version alignment across recommended tags
- Cross-image dependency specifier conflict detection
- File identity checks (de-vendor scripts, nginx configs)
- Tekton pipeline arg naming
- ImageStream kustomization consistency (tag order, placeholder patterns, env keys)
- `make --just-print` push-disable verification

[`tests/test_pylock_downgrade.py`](tests/test_pylock_downgrade.py):
- Compares current pylocks against git base ref to prevent version regressions

### 6b. Unit tests

[`tests/unit/scripts/test_pylocks_generator.py`](tests/unit/scripts/test_pylocks_generator.py):
- Tests for the lockfile generator: timestamp parsing, exclude-newer resolution, JSON format params
- Uses `repo_root` fixture and `tmp_path`

### 6c. Manual tests

[`tests/manual/`](tests/manual/) -- documented procedures (not automated):
- `runtime_elyra/README.md` -- step-by-step Elyra pipeline validation on RHOAI
- Manual GPU/PyTorch/TensorFlow test notebooks

---

## 7. Makefile `validate-*` targets (non-Jupyter image validation)

**Entry point:** `make validate-runtime-image`, `make validate-codeserver-image`, `make validate-rstudio-image`

These are **separate** from the `test-%` papermill flow. They assume a pod with a **fixed name** already exists (`runtime-pod`, `codeserver-pod`, `rstudio-pod`) and validate the image in-cluster via `kubectl exec`.

### 7a. `validate-runtime-image`

- `kubectl wait --for=condition=ready pod runtime-pod`
- Checks `which` for required CLI tools (`curl`, `python3`)
- Curls Elyra's `requirements-elyra.txt` from GitHub, `pip install`s it in-pod
- Curls **upstream** `simple_execute.ipynb` from nteract/papermill repo (not the repo's own test notebooks)
- Runs `python3 -m papermill simple_execute.ipynb output.ipynb` -- validates Elyra runtime compatibility

### 7b. `validate-codeserver-image`

- Binary-existence only: `which curl python oc code-server`
- No papermill, no notebook execution

### 7c. `validate-rstudio-image`

- `which` for R tools, `kubectl cp` a `test_script.R` into the pod
- Runs `R -e` for TinyTeX validation, `Rscript test_script.R`
- No papermill

**Key difference from pattern 1:** These use fixed pod names (not label selectors), test different image families (runtime/codeserver/rstudio vs Jupyter), and `validate-runtime-image` uses an **upstream** notebook rather than the repo's `test_notebook.ipynb`.

**Key file:** [`Makefile`](Makefile) (lines ~302-397)

---

## 8. Go unit tests

**Entry point:** `make test-unit` (Go portion) or `go test -C scripts/buildinputs -cover ./...`

**Location:** [`scripts/buildinputs/`](scripts/buildinputs/)
- [`buildinputs_test.go`](scripts/buildinputs/buildinputs_test.go) -- tests Dockerfile dependency parsing (`getDockerfileDeps`, bind mounts, referenced paths)
- [`heredoc_test.go`](scripts/buildinputs/heredoc_test.go) -- tests heredoc/`Doc()` string handling

**CI:** `.github/workflows/code-quality.yaml` runs **gotestsum** with coverage and JUnit output, uploads to Codecov.

No other Go test files exist in the repo (`scripts/check-payload/` has Go code but no tests).

---

## 9. CI linting and static analysis (non-pytest)

**Entry point:** `.github/workflows/code-quality.yaml` `code-static-analysis` job + `prek` job

These are **not** pytest-based; they run standalone tools:

- **yamllint** -- validates `*.yaml`/`*.yml` using [`ci/yamllint-config.yaml`](ci/yamllint-config.yaml)
- **hadolint** -- lints `Dockerfile*` using [`ci/hadolint-config.yaml`](ci/hadolint-config.yaml)
- **kustomize regression** -- [`ci/kustomize.sh`](ci/kustomize.sh): renders kustomize with two versions (5.0.3 vs 5.6.0), asserts identical output, no `_PLACEHOLDER` leftover, empty stderr
- **JSON-in-YAML validation** -- [`ci/check-json.sh`](ci/check-json.sh): uses `yq` + `json_verify` (yajl)
- **JSON schema validation** -- [`ci/validate_json.py`](ci/validate_json.py)
- **Codegen drift** -- [`ci/generate_code.sh`](ci/generate_code.sh): re-runs generators, fails if `git status` is dirty
- **prek (pre-commit)** -- [`.pre-commit-config.yaml`](.pre-commit-config.yaml): `ruff check` + `ruff format` (scoped to `ci/` and `tests/`), `pyright --pythonversion 3.14`, `uv lock --locked`, standard hooks (large files, TOML, merge conflicts)
- **pinact** -- SHA-pin check + verify for GitHub Actions

**Key file:** [`scripts/check_dockerfile_alignment.sh`](scripts/check_dockerfile_alignment.sh) -- compares `Dockerfile.konflux.*` vs base Dockerfiles (run as part of `make test`)

---

## 10. Security and vulnerability scanning

Multiple channels, none running as part of `make test`:

### 10a. Trivy (filesystem + image)

- [`.github/workflows/security.yaml`](.github/workflows/security.yaml) -- Trivy FS scan, SARIF upload to GitHub Security tab
- [`.github/workflows/build-notebooks-TEMPLATE.yaml`](.github/workflows/build-notebooks-TEMPLATE.yaml) -- optional Trivy on built images (label-gated or scheduled)
- Composite action at [`.github/actions/trivy-scan-action`](.github/actions/trivy-scan-action) -- Podman + `aquasec/trivy` image
- Config: [`trivy.yaml`](trivy.yaml) -- custom pip file patterns, severity MEDIUM+, skip vendored dirs

### 10b. Quay vulnerability reporting

- [`ci/security-scan/quay_security_analysis.py`](ci/security-scan/quay_security_analysis.py) -- calls Quay security API per image digest, aggregates CVE severities, writes `security_scan_results.md`
- Scheduled workflow (`.github/workflows/sec-scan.yml`): every third Friday, opens PR with updated report

### 10c. Konflux/Tekton pipeline scanning

- **Clair** scan and **Syft** SBOM generation tasks in `.tekton/` pipelines
- **Snyk SAST** on selected pipelines (e.g., minimal runtime, TrustyAI)

### 10d. Manifest-vs-reality validation

- [`manifest_validation_test.py`](tests/containers/manifest_validation_test.py) -- pytest (marker `manifest_validation`), compares imagestream annotation claims against `pip list` inside containers and Quay Clair API responses
- Triggered by [`.github/workflows/software-versions.yaml`](.github/workflows/software-versions.yaml)

### 10e. Developer CVE tooling

- [`scripts/cve/`](scripts/cve/) -- Jira tracker creation, Syft scanning, SBOM analysis (manual/developer use, not CI)

---

## 11. Infrastructure smoke workflows

GitHub Actions workflows that test **CI infrastructure itself**, not notebook images:

- **`test-setup-uv.yaml`** -- verifies `uv` install, Python version matches `.python-version`, `uv sync --locked` works
- **`test-install-podman.yaml`** -- `podman version`, `podman run nginx` on port 8080
- **`test-trivy-scan-action.yaml`** -- exercises the Trivy composite action (image full, image library-only, filesystem)
- **`test-build-notebooks-template.yaml`** -- integration test of the reusable `build-notebooks-TEMPLATE.yaml` workflow
- **`test-provision-k8s.yaml`** -- smoke test for the cluster provisioning composite action
- **`build-browser-tests.yaml`** -- after building the Playwright image, `podman run ... --list --project=chromium` validates it

---

## Summary: Pattern comparison

- **Papermill on OpenShift (1):** Shell-driven, in-cluster `kubectl exec`, tests Jupyter notebook execution e2e. Assertion via grep on stderr.
- **Testcontainers (2):** Python pytest, local Docker/Podman, image started with `sleep infinity` or real entrypoint. Tests exec into containers.
- **Python injection (3):** Three sub-patterns for getting Python code into containers -- base64+exec(), inline `-c` strings, and file copy+exec. Used by both testcontainers and K8s tests.
- **K8s API (4):** Python `kubernetes` + `openshift-python-wrapper`, creates real OpenShift resources (Namespace, PVC, Deployment), uses `kubernetes.stream` for exec and port-forward.
- **Browser (5):** Playwright with Node testcontainers for code-server, `@kubernetes/client-node` for OpenShift console discovery.
- **Static/unit pytest (6):** Default pytest run, no containers needed. Policy enforcement on Dockerfiles, manifests, pylocks. Includes doctests in `ntb/strings.py`.
- **Makefile validate-* (7):** Fixed-pod-name in-cluster validation for runtime/codeserver/rstudio images. Runtime uses upstream papermill notebook; others use `which`/R scripts.
- **Go unit tests (8):** `go test` in `scripts/buildinputs/` for Dockerfile dependency parsing.
- **CI linting (9):** yamllint, hadolint, kustomize regression, JSON validation, ruff/pyright via prek, pinact for action SHAs.
- **Security scanning (10):** Trivy (FS + image), Quay CVE reporting, Konflux Clair/Syft/Snyk, manifest-vs-reality pytest, developer CVE tooling.
- **Infra smoke (11):** Workflows that test CI tooling itself (uv, Podman, Trivy action, build template, k8s provisioning, browser test image).

---

## Markers and pytest configuration

Registered markers in [`pytest.ini`](pytest.ini): `openshift`, `cuda`, `rocm`, `buildonlytest`, `manifest_validation`

Key flags: `--strict-markers`, `--capture=no`, `--tb=short`, `--doctest-modules`, coverage enabled

Container tests are excluded from default collection via `collect_ignore = ["containers"]` in [`tests/conftest.py`](tests/conftest.py)

Doctest modules collected from `ntb/` (only `ntb/strings.py` has `>>>` examples)

**Comment by jiridanek** [comment]
@coderabbitai this is a bit too much looks like, please critically review what's happening here

**Comment by jiridanek** [comment]
@coderabbitai write a better issue (as a comment) then, will you

**Comment by jiridanek** [comment]
@coderabbitai open standalone github issue for Reorganizing tests/test_main.py — worth doing but independent

**Comment by jiridanek** [comment]
@coderabbitai, what about the in-testcontainers and in-cluster (k8s lib) split; can we write one test that does not care how/where container image was started?

**Comment by jiridanek** [comment]
@coderabbitai investigate container_exec_with_stdin (interactive TTY, used for RStudio tests) has no clean WebSocket equivalent in kubernetes_utils today. That's a narrower problem — it affects only stdin-driven tests.
